[MAGE] GH login + so many aMAGEing updates

vincanger opened the pull request:
### Description

-  put generations behind GH login
-  created a user dashboard page to view user's apps
- check if user has starred our repo via GH api
- update meta og:title to include MageGPT and Mage GPT
- new twitter banner with updated name

![image](https://github.com/wasp-lang/wasp/assets/70215737/7037b62a-0577-4fd9-a2f4-baf6d6090f9c)
![image](https://github.com/wasp-lang/wasp/assets/70215737/9122c1a6-82fe-4c5f-8881-4079ba0d73f1)

### Select what type of change this PR introduces:

1. [ ] **Just code/docs improvement** (no functional change). 
2. [ ] **Bug fix** (non-breaking change which fixes an issue).
3. [x] **New feature** (non-breaking change which adds functionality).
4. [ ] **Breaking change** (fix or feature that would cause existing functionality to not work as expected).

infomiho commented on wasp-ai/src/client/pages/MainPage.jsx:
```diff
@@ -67,19 +76,40 @@ const MainPage = () => {
+  }, [userProjects, user]);
+
+  async function checkIfUserStarredWasp({ username }) {
+    const token = "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";
```
@vincanger This is big no-no 😢 we should do the check on the backend to keep the token a secret.

**You should delete the leaked token right away.**

vincanger replied:
yep. pushed the wrong commit too soon. I had it already refactored to be a query but forgot to commit the change before creating the PR

vincanger commented on wasp-ai/src/server/operations.ts:
```diff
@@ -222,7 +209,46 @@ export const getStats = (async (_args, context) => {
+    const response = await fetch(`https://api.github.com/user/starred/wasp-lang/wasp`, {
+      method: "GET",
+      headers: {
+        Authorization: "Basic " + btoa(username + ":" + process.env.GITHUB_PERSONAL_ACCESS_TOKEN),
```
here she is @infomiho

infomiho replied:
Nice job!

infomiho commented on wasp-ai/src/client/components/Header.jsx:
```diff
@@ -1,25 +1,58 @@
+              <RxQuestionMarkCircled className="text-base text-slate-600" />
+            </a>
+
+            <div className="relative group">
```
I suggest adding `flex items-center`

<img width="338" alt="Screenshot 2023-10-30 at 11 38 08" src="https://github.com/wasp-lang/wasp/assets/2223680/3065123f-d7e8-43c2-aee6-c7f629aa889e">
<img width="305" alt="Screenshot 2023-10-30 at 11 39 48" src="https://github.com/wasp-lang/wasp/assets/2223680/2c5c31bc-13aa-4000-bfd3-2e9adb731659">

infomiho commented on wasp-ai/src/client/pages/UserPage.jsx:
```diff
@@ -0,0 +1,101 @@
+              </tr>
+            </thead>
+            <tbody>
+              {projects?.length &&
```
<img width="674" alt="Screenshot 2023-10-30 at 11 44 58" src="https://github.com/wasp-lang/wasp/assets/2223680/7755d1ad-3c84-435c-871e-17796f714ea6">

To avoid this always use `.length > 0`

Martinsos replied:
Interesting, I would expect 0 to be false! Why would he need to add > 0?

vincanger replied:
yeah I assumed the same, but seems this is how react behaves...

Martinsos replied:
Aha, ok so what happens is following:

You have `project?.length && projects.map(...)`.

So, if the first part of `&&` is "falsy", e.g. `undefined` or `0`, then that is the value of the whole expression, and what React gets to render is `undefined` or `0`.

On the other hand, if it is "truthy", e.g. `1` or something, then it will render that `project.map(...)`.

So what happens when React gets `undefined` or `0` to render? Well, if it gets `undefined`, I guess it doesn't render anything, but if it gets `0`, it renders `0`! So that is the issue here.

What @infomiho wrote will help because it will ensure the first part of `&&` is boolean: True or False (since `>` is boolean operator). And React, when given `False`, doesn't render anything, so that is good.

You could also do `!!(project?.length)`.

Martinsos commented on wasp-ai/src/client/pages/StatsPage.jsx:
```diff
@@ -337,11 +337,11 @@ function getWaitingInQueueDuration(stat, logsByProjectId) {
 }
 
-function getColorValue(colorName) {
+export function getColorValue(colorName) {
```
How is it that these got exported?

If they are used somewhere else, maybe it would make sense to extract them into separate file, because why would they be in StatsPage.jsx?

Martinsos commented on wasp-ai/src/client/pages/UserPage.jsx:
```diff
@@ -0,0 +1,101 @@
+export function UserPage({ user }) {
+  const { data: projects, isLoading, error } = useQuery(getProjectsByUser);
+
+  if (isLoading) "Loading Projects...";
```
Hm hoe does this work? Does this work? I would expect these two lines to do nothing at all actually, they have string, which is an expression, playing the role of the statement, which means nothing happens.

It is like writing
```js
console.log('Hi');
'hellou'
```

Basicaly this second line gets evaluated and then value is just dropped.

Martinsos replied:
I don't think it is, that is not how JS works.

These lines are basically no-op -> they don't do anything, you could remove them from you program and its functioning won't change.

Try this:
```js
function vince (x) {
    if (x === 1) "one";
    if (x === 2) "two";
    return "not one or two";
}
```

Whichever argument you give it, it will return `"not one or two"`.

JS program (and same is for most other languages, like C++, python, Java, ...) is a list of statements, at the top level. So a bunch of statements:

STMT
STMT
STMT

Statement can be something like `if () {}` or `for ...` or `import ...` or `function foo () {...}` and so on.
Then we also have expressions, which are stuff like `1` or `1 + 2` or `vince(3)` and so on.

Statements usually have expressions in them as their parts.

But, just a standalone expression also counts as statement!

That makes sense, because that allows you to call a function alone in its line, because function call is an expression.
But it also allows stuff like:

```js
console.log("hi");

"hi";

return 3;
```

What is that "hi" doing there on its own? Well, nothing really. It is a statement that is really just an expressions (because an expression is also a valid statement), and interpreter / compiler will go line by line, evaluate console.log("hi") and write "hi" to stdout, then it will evaluate "hi" and say "ok, this is 'hi', but nothing to do with it so I will just leave it here", and then `return 3` ends the function and so on.

```
if (...) "hi";
```

will have the same effect as just
```
"hi"
```

which is nothing, it will just execute only sometimes.

If somehow possibly I got this wrong, I apologize, but I am 99.99% confident in what I am writing here. I am leaving 0.01% because you never know.

vincanger replied:
whoop! i forgot to add `return`

so it should be 
```
if (isLoading) return "Loading..."
```

Martinsos replied:
Damn :D! Yes that is fine :D.

Martinsos commented on wasp-ai/src/client/pages/UserPage.jsx:
```diff
@@ -0,0 +1,101 @@
+            <tbody>
+              {projects?.length &&
+                projects.map((project) => (
+                  <tr className="bg-white border-t" key={project.id}>
```
I would extract this as a sub-component.

Martinsos commented on wasp-ai/src/server/operations.ts:
```diff
@@ -222,7 +209,46 @@ export const getStats = (async (_args, context) => {
+  });
+};
+
+import { CheckIfUserStarredWasp } from "@wasp/queries/types";
```
Ok wild, import in the middle of the document :D. Maybe we should put it at the top of the document?

Martinsos commented on wasp-ai/src/server/operations.ts:
```diff
@@ -222,7 +209,46 @@ export const getStats = (async (_args, context) => {
+    console.error("Error:", error);
+  } finally {
+    return status;
+  }
```
So if error happens, you will return `false`, not an error.
Is that ok? Wouldn't we rather return an error in that case?
And then on frontend, if we get error, we would rather not bother them with leaving a star? Because right now, let's say we breach some rate limit or something, and we get all errors, then we will be bothering people who even starred.

Martinsos replied:
Is this comment irrelevant now?

vincanger replied:
yes we completely removed this function due to the GH API issue

Martinsos commented on wasp-ai/src/server/operations.ts:
```diff
@@ -222,7 +209,46 @@ export const getStats = (async (_args, context) => {
+    const response = await fetch(`https://api.github.com/user/starred/wasp-lang/wasp`, {
+      method: "GET",
+      headers: {
+        Authorization: "Basic " + btoa(username + ":" + process.env.GITHUB_PERSONAL_ACCESS_TOKEN),
```
So we will make this check every time user comes to the Mage? I wonder if we are going to hit some Github rate limits. Could we maybe cache it in some way? For example, if we realized they starred us, we store that into our database and don't ask them again? Yeah, maybe that would be a good way to do it? Sure, they could unstar us, but that is really rare and I don't we want to bother them in that case anyway, I would let them be.

Martinsos replied:
Ok this is irrevelant now.

Martinsos reviewed (commented):
Awesome job @vincanger !

I think all is pretty solid, and nice job on doing all these changes! I left some comments on how to improve things, check them out.

Martinsos commented:
> Hello from my test account 👋
> 
> It seems that the method for checking if a user starred the repo doesn't work 😢

Oh wow, why not?

vincanger commented on wasp-ai/src/client/components/Header.jsx:
```diff
@@ -1,25 +1,58 @@
+
+            <div className="relative group">
+              <button
+                onClick={() => {
+                  if (!user) {
+                    setIsLoginModalOpen(true);
+                  } else {
+                   window.location.href = "/user";
+                  }
+                }}
```
1. I avoid sending them to the `/login` route as that's reserved for admins/stats. We could set up two seperate login routes though...

2. yeah I'll use history for this one

Martinsos replied:
Aha interesting, I forgot about that!

So we have users log in with github, and we log in with gmail, is that so? Interesting!

So what happens if I try to access my user profile or something un-authenticated, I get thrown to home page?

Oh, question -> if I get a result of generation and I share it with somebody, do they need to authenticate to see the result?

vincanger replied:
1. yep

2. no, results page is unauthenticated for that reason.

vincanger commented on wasp-ai/src/client/components/Header.jsx:
```diff
@@ -1,25 +1,58 @@
+          </StatusPill>
+        </div>
+      )}
+      {isUserPage && (
```
makes sense! hadn't thought about the dependency injection for this case. adding that now!

vincanger commented on wasp-ai/src/client/pages/MainPage.jsx:
```diff
@@ -95,19 +108,23 @@ const MainPage = () => {
+          appCreativityLevel: creativityLevel.value,
+        })
+      );
+      if (!user) {
+        setIsLoginModalOpen(true);
+        return;
+      }
```
yes. this way whenever they navigate away from the main page and come back, their previous prompt/description is there again for them to work with. this also makes it easier for them to iterate on their app prompt with each generatiom.

Martinsos replied:
Ok, nice DX!

vincanger commented on wasp-ai/src/client/pages/MainPage.jsx:
```diff
@@ -95,19 +108,23 @@ const MainPage = () => {
+        return;
+      }
     } catch (error) {
-      console.error(error)
+      console.error(error);
     }
```
miho said localStorage methods often fails (quota exceeded, browser security restriction) so it's good practice to wrap it in a try-catch block to prevent it from crashing the app.

Martinsos replied:
Oh wow had no idea about that! Right, if that is so then it makes sense.

vincanger commented on wasp-ai/src/client/pages/ResultPage.jsx:
```diff
@@ -73,10 +74,16 @@ export const ResultPage = () => {
   useEffect(() => {
-    if (currentStatus.status === "success") {
-      setIsStarRepoOpen(true);
+    const currentAppStatus = appGenerationResult?.project?.status;
+    if (currentAppStatus === "in-progress") {
+      setIsRunning(true)
+    } else if (currentAppStatus === "success" && isRunning) {
+      setIsSuccessModalOpen(true)
+      setIsRunning(false)
+    } else if (isRunning) {
+      setIsRunning(false)
     }
```
yeah why didn't I think of that :)

vincanger commented:
> Awesome job @vincanger !
> 
> I think all is pretty solid, and nice job on doing all these changes! I left some comments on how to improve things, check them out.

I addressed all your comments and tried to clean up and extract as many things as I could. I think the header works a lot nicer now that it accepts children and the StatusPill as a dependency.

To address the GH api issue, we need to send the API the user's access token which we receive on successful login to GH via OAuth. Unfortunately, Wasp doesn't expose this token in any way for us to use, so that is out of the question.

Martinsos commented on wasp-ai/src/client/pages/MainPage.jsx:
```diff
@@ -67,19 +82,17 @@ const MainPage = () => {
   const [appAuthMethod, setAppAuthMethod] = useState(availableAuthMethods[0]);
 
+  useEffect(() => {
+    if (!hasUserStarredWasp && userProjects?.length >= 1) {
```
Ok!

Martinsos commented:
> > Awesome job @vincanger !
> > I think all is pretty solid, and nice job on doing all these changes! I left some comments on how to improve things, check them out.
> 
> I addressed all your comments and tried to clean up and extract as many things as I could. I think the header works a lot nicer now that it accepts children and the StatusPill as a dependency.
> 
> To address the GH api issue, we need to send the API the user's access token which we receive on successful login to GH via OAuth. Unfortunately, Wasp doesn't expose this token in any way for us to use, so that is out of the question.

Oh that is interesting. Is that a missing feature in Wasp? Sounds like it is something we should add? @infomiho you are the expert here, what are your thoughts? Should we create an GH issue for it, if we don't already have one?

vincanger commented:
ok @Martinsos, I added the return declaration to the UserPage, pushed the changes, and addressed all remaining comments.

infomiho commented:
Yes, we are not giving the developer access to the Github Access Token in the `getUserFields` function therefore they can't do API calls on the behalf of the user.

https://github.com/wasp-lang/wasp/blob/main/waspc/data/Generator/templates/server/src/auth/providers/oauth/init.ts#L46

It should probably be an issue, but I wouldn't add it "hey we need to support giving the access token to the user" I would do it inside of the overall auth effort of moving to https://github.com/wasp-lang/openid-client-experiment and see how we do it.

Martinsos commented on wasp-ai/src/client/pages/UserPage.jsx:
```diff
@@ -0,0 +1,107 @@
+                <span className="flex gap-1">
+                  {project.zipDownloadedAt && (
+                    <span
+                      title={`Downlaoded ${format(project.zipDownloadedAt)}`}
```
```suggestion
                      title={`Downloaded ${format(project.zipDownloadedAt)}`}
```
Let's fix this while at it.

Martinsos reviewed (commented):
@vincanger I did another round of reviewing!

vincanger commented on wasp-ai/src/client/stats/stats.js:
```diff
@@ -0,0 +1,37 @@
+import { availableColors } from "../components/Color";
+
```
would `display/displayHelpers.js` be better?

Martinsos replied:
I have to admit I am not 100% sure what they are about! We need to figure out some names here :). But `display/utils.js` doesn't sound so bad at the first glance.

All of these functions need richer names, if not comments.

vincanger replied:
ok yeah good point. I re-used these functions from the stats page for the user page without really knowing what they did, but I think I came up with good names for them now :)

Martinsos replied:
Nice, I will close this comment as I continued commenting on each one of them individually!

vincanger commented on wasp-ai/src/client/pages/UserPage.jsx:
```diff
@@ -0,0 +1,107 @@
+  );
+}
+
+function UserTable({ projects }) {
```
good idea. 

...and I thought I was getting better at naming 🤦‍♂️

Martinsos replied:
I am sure you are! It is about the journey, not the destination :P.

vincanger commented:
ok @Martinsos -- reviewd and updated. awaiting your final review :)

vincanger commented:
@Martinsos I feel pretty confident that the naming of names was done in the name of clean code :)

Martinsos commented on wasp-ai/src/client/project/converters.js:
```diff
@@ -19,7 +21,8 @@ export function getStatusName(status) {
 
-export function getStatusText(status) {
+// converts project's status to displayable text
+export function convertProjectStatusToText(status) {
```
Ok, this one is not bad. Maybe I would go with `projectStatusToDisplayableText` or `projectStatusToPrettyText` and then you can ditch the comment.

vincanger replied:
sounds good, but part of me now wants to use the same `getXforY` naming structure as the other util functions to be consistent. something like e.g. `getDisplayableTextForProjectStatus`

Martinsos replied:
Yup that also sounds good, why not.

There is a slight difference I think: Here it feels like conversion is 1 to 1 -> you are not loosing any info really, you are converting from ugly to nice string, from id to string.

On the other functions, you are getting a style (class name) for a concept. Project color -> style. Ok, that sounds 1 on 1. But what about project status to style? There is information loss there. Maybe different statuses have same styles. It would certainly be weird to go back from style to status. Maybe also from style to color name, although less weird, depends on the situation.

So getXforY sounds good to me when there is information loss in this "conversion", because then it is not really a conversion, it is instead "describe me this thing (Y) from this perspective (X)" aka "give me X for Y".

Btw this is all very subtle and whatever you do would be ok really, be it `getXForY` or `XtoY`, but this is kind of reasoning I used while deciding what to use.

Martinsos commented on wasp-ai/src/client/project/converters.js:
```diff
@@ -1,10 +1,12 @@
 import { availableColors } from "../components/Color";
 
-export function getColorValue(colorName) {
+// converts project's color name to tailwind css class name
```
Extra thought: Ok, you could potentially drop `Tailwind` part of the name, if you wanted to be agnostic of the styling library/framework being used. I think it makes everything a tiny bit more vague, but it is not terrible. A comment would help in that case. I would still go with "Tailwind" though.

vincanger replied:
`availableColors`, and a lot of the other names, were already there when I started working on the GH login, so I didn't think to change them. but this definitely makes a lot of sense and will keep this in mind for the future.

Martinsos commented on wasp-ai/src/client/pages/StatsPage.jsx:
```diff
@@ -14,7 +14,11 @@ import { Header } from "../components/Header";
 import { HomeButton } from "../components/Header";
-import { getColorValue, getStatusName, getStatusText } from "../display/utils"
+import {
+  convertProjectColorToClassName,
```
If these names are too long to be comfortably used in this file, you can shorten them with `as`. You could do `convertProjectColorToClassName as getColorClassName`. It is localized in this file, so no biggie, and if somebody is not very clear as to what `getColorClassName` really does, they can look it up here and figure it out.

Martinsos commented on wasp-ai/src/client/components/Header.jsx:
```diff
@@ -1,25 +1,73 @@
-      )}
+      <div className="flex flex-col items-end gap-2">
+        <div className="flex items-center gap-3 m-1">{children}</div>
+        {!!currentStatus && (
+          <StatusPill status={currentStatus.status} className="hidden md:flex">
+            {currentStatus.message}
+          </StatusPill>
```
Nice

Martinsos commented on wasp-ai/src/client/pages/MainPage.jsx:
```diff
@@ -67,19 +76,17 @@ const MainPage = () => {
   const [appAuthMethod, setAppAuthMethod] = useState(availableAuthMethods[0]);
 
+  useEffect(() => {
+    if (userProjects?.length > 0 && userProjects.length % 2 === 0) {
```
Ok this is a bit peculiar :D, so what are we doing, we are asking them for stars on every second project? Even if they left a start already? Before we were asking on the second project, but now we are asking for every second project? Is this maybe a mistake (ChatGPT did you do this)?

vincanger replied:
no I did this, because we have no way of knowing if they starred us or not yet, and I'm assuming most people won't create more than 2 or 3 projects. If they do, these are the people we need to annoy :)

Martinsos replied:
Ah got it! Ok sorry for assuming too much! No worries then, do as you think is best!

Martinsos reviewed (changes requested):
@vincanger nice job with cleaning up the names -> we are not yet there completely I think, but you brought them close enough so I could understand what they are, which is already great! I left some suggestions on how to possibly improve them further + a few other small comments and one possible mistake.

Please go through all of these, also resolve any comments where my response was last and you are happy with it, and then I will take another look!

vincanger commented:
ok @Martinsos ! the renaming of renamed names is complete :)

Martinsos commented on wasp-ai/src/client/project/utils.js:
```diff
@@ -0,0 +1,40 @@
+import { validProjectBrandColors } from "../components/Color";
+
+// converts project's color name to tailwind css class name
```
This comment is redundant now.

Martinsos commented on wasp-ai/src/client/project/utils.js:
```diff
@@ -0,0 +1,40 @@
+  return validProjectBrandColors.find((color) => color.name === colorName).color;
+}
+
+// converts project's status to tailwind css class name
```
This comment is also redundant now.

Martinsos commented on wasp-ai/src/client/project/utils.js:
```diff
@@ -0,0 +1,40 @@
+  }
+}
+
+// converts project's status to displayable text
```
Also redundant comment!

Martinsos reviewed (approved):
@vincanger nice job, I think this is it!

I left a couple small comments, please give them a look and handle them as you wish, and then you can merge! I approved the PR.
